### PR TITLE
PARQUET-287: Keep a least 1 column from union members when projecting thrift unions

### PR DESCRIPTION
**Pull request by isnotinvain**
Currently, the projection API allows you to select only some "kinds" of a union, or to drop a required union entirely. This becomes an issue when assembling these records, as they will be appear to be unions of an unknown type (how do you coerce an empty struct into a union?). The way this case is handled for primitives is by supplying a default value (like 0, or null). However, with a union, you have to choose what "kind" of the union it will act as, and in the interest of not being misleading, this PR reads one column to figure out what the correct "kind" is.

In the future, the better solution is to filter these records out -- a projection is really a request for a filter in this case. But for now, this should get us correctness without involving the filter API.

I think this PR needs some more tests before merging, but I wanted to get it out and get some feedback now. 

I also refactored how ThriftSchemaVisitor works to not be stateful, by explicitly passing state through the recursion -- this makes it much easier to reason about.

*edit* This PR also includes a fix for PARQUET-275 because I encountered it during testing.

*edit 2* This PR also includes a fix for PARQUET-283

### CODE REVIEW

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 1]
```diff
@@ -55,203 +73,239 @@
...
+    ThriftField keyField = mapType.getKey();
+    ThriftField valueField = mapType.getValue();
+
+    ConvertedField convKey = keyField.getType().accept(this,
+        new State(state.path.push(keyField), REQUIRED, "key"));
+
+    ConvertedField convValue = valueField.getType().accept(this,
+        new State(state.path.push(valueField), OPTIONAL, "value"));
+
+    if (!convKey.isKeep()) {
```
Comment (isnotinvain) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: Note: add tests for `Map<Union, T>`, `Map<T, Union>`, `Map<Union, Union>`, `List<Union>` etc.

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 2]
```diff
@@ -18,36 +18,54 @@
  */
```
Comment (isnotinvain) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: this file changed enough that you might just want to hit "view" above and read the new copy of it

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 3]
```diff
@@ -55,203 +73,266 @@
...
+    State state = new State(new FieldsPath(), REPEATED, "ParquetSchema");
 
-  public ThriftSchemaConvertVisitor(FieldProjectionFilter fieldProjectionFilter) {
-    this.fieldProjectionFilter = Preconditions.checkNotNull(fieldProjectionFilter, "fieldProjectionFilter");
+    ConvertedField conv = struct.accept(new ThriftSchemaConvertVisitor(filter, true), state);
+
+    MessageType messageType;
+
+    if (!conv.isKeep()) {
+      // TODO: should this be fatal? Projecting away everything?
```
Comment (isnotinvain) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: this should be fatal, right?

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ConvertedField.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+ */
+package org.apache.parquet.thrift;
+
+import org.apache.parquet.ShouldNeverHappenException;
+import org.apache.parquet.schema.Type;
+import org.apache.parquet.thrift.projection.FieldsPath;
+
+import static org.apache.parquet.Preconditions.checkNotNull;
+
+public interface ConvertedField {
```
Comment (tsdeng) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ConvertedField.java]: Java doc on the responsibility and usages of this interface?

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java (hunk 1 above)
Comment (isnotinvain) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: fixed

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java (hunk 3 above)
Comment (isnotinvain) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: @tsdeng what do you think?

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 5]
```diff
@@ -55,203 +73,266 @@
...
 
-  private Type currentType;
-  private Type.Repetition currentRepetition = Type.Repetition.REPEATED; // MessageType is repeated GroupType
-  private String currentName = "ParquetSchema";
+  public static MessageType convert(StructType struct, FieldProjectionFilter filter) {
+    State state = new State(new FieldsPath(), REPEATED, "ParquetSchema");
 
-  public ThriftSchemaConvertVisitor(FieldProjectionFilter fieldProjectionFilter) {
-    this.fieldProjectionFilter = Preconditions.checkNotNull(fieldProjectionFilter, "fieldProjectionFilter");
+    ConvertedField conv = struct.accept(new ThriftSchemaConvertVisitor(filter, true), state);
```
Comment (tsdeng) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: no need to save typing here. conv can mean converter and other things. convertedField is fine

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java (hunk 3 above)
Comment (tsdeng) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: Should be fatal, but in this PR https://github.com/apache/parquet-mr/pull/195
It enforces GroupType to have at least one child.. But may be throwing here in your code is more clear for debugging purposes?

**Comment by tsdeng** [comment]
Minor comments, otherwise LGTM

**Comment by tsdeng** [comment]
+1

**Pull request closed without merging**